enarx-keep-sev-shim/src/frame_allocator.rs

```diff
@@ -0,0 +1,192 @@
+    pub unsafe fn init(&mut self, boot_info: &BootInfo) {
+        let start = Address::from(boot_info.code.end);
+        self.max = boot_info.mem_size;
+        self.next_4k = start.raise();
+        self.next_2M = start.raise();
+    }
+}
+
+unsafe impl FrameAllocator<Size4KiB> for BootInfoFrameAllocator {
+    fn allocate_frame(&mut self) -> Option<PhysFrame> {
+        let frame_address = self.next_4k;
+        self.next_4k += Offset::from_items(1);
+
+        // if at the start of a 2MiB page, advance to the next free 2MiB
+        if self.next_4k.raw() == self.next_4k.raise::<Page2MiB>().raw() {
+            self.next_4k = self.next_2M.raise();
+        }
+
+        // we have taken a bite out of the next 2MiB page
+        // so advance the 2MiB pointer
+        if frame_address.raw() >= self.next_2M.raw() {
+            self.next_2M += Offset::from_items(1);
+        }
+
+        if self.next_4k.raw() > self.max {
+            return None;
+        }
+
+        Some(PhysFrame::containing_address(PhysAddr::new(
+            frame_address.raw() as _,
+        )))
+    }
+}
+
+unsafe impl FrameAllocator<Size2MiB> for BootInfoFrameAllocator {
+    fn allocate_frame(&mut self) -> Option<PhysFrame<Size2MiB>> {
+        let frame_address = self.next_2M;
+        self.next_2M += Offset::from_items(1);
+
+        if self.next_2M.raw() > self.max {
+            return None;
+        }
+
+        Some(PhysFrame::containing_address(PhysAddr::new(
+            frame_address.raw() as _,
+        )))
+    }
+}
+
+/// Allocate memory and map it to the given virtual address
+pub fn allocate_and_map_memory(
+    map_to: VirtAddr,
+    size: usize,
+    mapper: &mut OffsetPageTable,
+    frame_allocator: &mut BootInfoFrameAllocator,
+) -> Result<&'static mut [u8], ()> {
+    #[inline(always)]
+    pub fn do_allocate_and_map_memory<S: PageSize>(
+        map_to: &[Page<S>],
+        mapper: &mut impl Mapper<S>,
+        frame_allocator: &mut (impl FrameAllocator<S> + FrameAllocator<Size4KiB>),
+    ) -> Result<(), ()> {
+        if map_to.len() == 0 {
+            return Ok(());
+        }
+        let size = map_to.len() * (Page::<S>::SIZE as usize);
+
+        let page_range = {
+            let heap_start = VirtAddr::from_ptr(map_to.as_ptr());
+            let heap_end = heap_start + size - 1u64;
+            let heap_start_page = Page::<S>::containing_address(heap_start);
+            let heap_end_page = Page::<S>::containing_address(heap_end);
+            Page::range_inclusive(heap_start_page, heap_end_page)
+        };
+
+        for page in page_range {
+            let frame = frame_allocator.allocate_frame().ok_or(())?;
+
+            let flags = PageTableFlags::PRESENT | PageTableFlags::WRITABLE;
+            unsafe {
+                mapper
+                    .map_to(page, frame, flags, frame_allocator)
+                    .map_err(|_| ())?
+                    .flush();
+            }
+        }
+        Ok(())
+    }
```

Comment:
I think the same kind of accessibility/namespacing you're going for here can be accomplished by putting `do_allocate_and_map_memory` and `allocate_and_map_memory` into a submodule and only making `allocate_and_map_memory` `pub`. I think doing it that way is a bit easier on the eyes.